sys/libkern/strlen.c

```diff
@@ -95,19 +94,34 @@ size_t
 	 * p and (p & ~LONGPTR_MASK) must be equally accessible since
 	 * they always fall in the same memory page, as long as page
 	 * boundaries is integral multiple of word size.
+         *
+         * This is not true for CHERI, so we skip directly to byte
+	 * access if not word-aligned.
 	 */
-	lp = (const unsigned long *)((uintptr_t)str & ~LONGPTR_MASK);
+	lp = (const unsigned long *)rounddown2(str, sizeof(long));
+#ifdef __CHERI_PURE_CAPABILITY__
+	byte_check = (lp < str);
+	if (byte_check)
+		lp++;
+#else
 	va = (*lp - mask01);
 	vb = ((~*lp) & mask80);
 	lp++;
-	if (va & vb)
+	byte_check = (bool)(va & vb);
+#endif
+	if (byte_check)
 		/* Check if we have \0 in the first part */
 		for (p = str; p < (const char *)lp; p++)
 			if (*p == '\0')
 				return (p - str);
 
 	/* Scan the rest of the string using word sized operation */
-	for (; ; lp++) {
+#ifdef __CHERI_PURE_CAPABILITY__
+	for (; cheri_getlen(lp) - cheri_getoffset(lp) >= sizeof(long); lp++)
```

Comment:
There is also `cheri_bytes_remaining()` in cheric.h. However, that set negative values to zero so might add extra overhead. I think trapping if you pass OOB caps is fine so we probably don't need that.

Maybe we want a  variant of that helper that can return negative values?

Reply:
Or an instruction that gives the remaining addressable bytes.

Reply:
I'm not sure about adding a `cheri_bytes_remaining` variant just for one use case. I'd leave the explicit subtraction here as we avoid the extra check for zeroing when OOB and feels readable enough.